File: web/src/lib/bridge/scenarioReports.ts
// Automation scenario run reports: Tauri → app_data_dir/scenario_reports; browser → localStorage (keeping the latest 30).
import type { ScenarioReportSummary, ScenarioRunRecord } from "@/data/types";
import { isTauri, tauriInvoke } from "./client";

const LS_KEY = "orbit.scenarioReports";
/** Browser channel capacity cap (localStorage is limited to 5MB and reports keep growing) */
const MAX_WEB_REPORTS = 30;

function readWebReports(): ScenarioRunRecord[] {
  try {
    const raw = localStorage.getItem(LS_KEY);
    if (!raw) return [];
    const parsed = JSON.parse(raw);
    return Array.isArray(parsed) ? (parsed as ScenarioRunRecord[]) : [];
  } catch (e) {
    console.error("[scenarioReports] failed to read local reports", e);
    return [];
  }
}

function writeWebReports(list: ScenarioRunRecord[]): void {
  try {
    localStorage.setItem(
      LS_KEY,
      JSON.stringify(list.slice(0, MAX_WEB_REPORTS)),
    );
  } catch (e) {
    console.error("[scenarioReports] failed to write local reports", e);
  }
}

export function toReportSummary(r: ScenarioRunRecord): ScenarioReportSummary {
  return {
    id: r.id,
    workspaceId: r.workspaceId,
    targetType: r.targetType,
    targetId: r.targetId,
    targetName: r.targetName,
    runMode: r.runMode,
    startedAt: r.startedAt,
    durationMs: r.durationMs,
    envName: r.envName,
    status: r.status,
    totalPass: r.totalPass,
    totalFail: r.totalFail,
    totalSkip: r.totalSkip,
    caseCount: r.cases.length,
  };
}

/** Save a run report */
export async function saveScenarioReport(
  record: ScenarioRunRecord,
): Promise<void> {
  if (isTauri()) {
    await tauriInvoke("save_scenario_report", {
      req: {
        report: record as unknown as Record<string, unknown>,
        workspace_id: record.workspaceId ?? null,
      },
    });
    return;
  }
  const list = readWebReports().filter((r) => r.id !== record.id);
  list.unshift(record);
  writeWebReports(list);
}

/** List report summaries (newest startedAt first) */
export async function listScenarioReports(
  workspaceId?: string | null,
): Promise<ScenarioReportSummary[]> {
  if (isTauri()) {
    const rows = await tauriInvoke<ScenarioReportSummary[]>(
      "list_scenario_reports",
      { workspaceId: workspaceId ?? null },
    );
    return rows ?? [];
  }
  return readWebReports()
    .filter((r) => !workspaceId || (r.workspaceId ?? null) === workspaceId)
    .map(toReportSummary)
    .sort((a, b) => b.startedAt - a.startedAt);
}

/** Load report details */
export async function loadScenarioReport(
  id: string,
): Promise<ScenarioRunRecord | null> {
  if (isTauri()) {
    return tauriInvoke<ScenarioRunRecord>("load_scenario_report", { id });
  }
  return readWebReports().find((r) => r.id === id) ?? null;
}

/** Delete a report */
export async function deleteScenarioReport(id: string): Promise<void> {
  if (isTauri()) {
    await tauriInvoke("delete_scenario_report", { id });
    return;
  }
  writeWebReports(readWebReports().filter((r) => r.id !== id));
}

/** Clear all reports */
export async function clearScenarioReports(): Promise<void> {
  if (isTauri()) {
    await tauriInvoke("clear_scenario_reports");
    return;
  }
  writeWebReports([]);
}

File: web/src-tauri/src/commands/scenario_reports.rs
//! Tauri commands: persist automation scenario run reports (app_data_dir/scenario_reports/<id>.json).
//!
//! Kept isolated from load-test reports (orbit_server::reports, ~/.orbit/reports):
//! scenario reports are aggregated by the frontend; this only does file CRUD + capacity trimming (keeps the newest MAX_KEEP entries).

use serde::Deserialize;
use serde_json::Value;
use std::path::PathBuf;
use tauri::Manager;

/// Number of newest reports to keep (oldest evicted in ascending startedAt order once exceeded)
const MAX_KEEP: usize = 200;

fn reports_dir(app: &tauri::AppHandle) -> Result<PathBuf, String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("failed to get data dir: {}", e))?;
    Ok(dir.join("scenario_reports"))
}

#[derive(Debug, Deserialize)]
pub struct SaveScenarioReportRequest {
    /// Full report object (serialized frontend ScenarioRunRecord)
    pub report: Value,
    #[serde(default)]
    pub workspace_id: Option<String>,
}

/// Save a run report
#[tauri::command]
pub fn save_scenario_report(
    app: tauri::AppHandle,
    req: SaveScenarioReportRequest,
) -> Result<Value, String> {
    let id = req
        .report
        .get("id")
        .and_then(|v| v.as_str())
        .ok_or_else(|| "report is missing the id field".to_string())?
        .to_string();
    let dir = reports_dir(&app)?;
    std::fs::create_dir_all(&dir).map_err(|e| format!("failed to create reports dir: {}", e))?;

    let mut report = req.report;
    if report.get("workspaceId").is_none() {
        if let Some(ws) = req.workspace_id {
            report["workspaceId"] = Value::String(ws);
        }
    }
    let body = serde_json::to_string_pretty(&report).map_err(|e| e.to_string())?;
    let path = dir.join(format!("{}.json", id));
    let tmp = dir.join(format!("{}.json.tmp", id));
    std::fs::write(&tmp, body.as_bytes()).map_err(|e| format!("failed to write report: {}", e))?;
    // On Windows, rename cannot overwrite an existing file, so delete first
    if path.exists() {
        std::fs::remove_file(&path).map_err(|e| format!("failed to replace old report: {}", e))?;
    }
    std::fs::rename(&tmp, &path).map_err(|e| format!("failed to save report: {}", e))?;

    trim_reports(&dir)?;
    Ok(serde_json::json!({ "status": "ok", "id": id }))
}

/// List reports (newest startedAt first); empty workspace_id = all
#[tauri::command]
pub fn list_scenario_reports(
    app: tauri::AppHandle,
    workspace_id: Option<String>,
) -> Result<Value, String> {
    let dir = reports_dir(&app)?;
    if !dir.exists() {
        return Ok(Value::Array(vec![]));
    }
    let mut items: Vec<Value> = Vec::new();
    for entry in std::fs::read_dir(&dir).map_err(|e| e.to_string())? {
        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        let content = match std::fs::read_to_string(&path) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let value: Value = match serde_json::from_str(&content) {
            Ok(v) => v,
            Err(_) => continue,
        };
        if let Some(ws) = &workspace_id {
            let own = value.get("workspaceId").and_then(|v| v.as_str());
            if own != Some(ws.as_str()) {
                continue;
            }
        }
        items.push(value);
    }
    items.sort_by(|a, b| {
        let ta = a.get("startedAt").and_then(|v| v.as_i64()).unwrap_or(0);
        let tb = b.get("startedAt").and_then(|v| v.as_i64()).unwrap_or(0);
        tb.cmp(&ta)
    });
    // The list returns only summary fields to avoid reading all details at once
    let summaries: Vec<Value> = items
        .iter()
        .map(|r| {
            serde_json::json!({
                "id": r.get("id"),
                "workspaceId": r.get("workspaceId"),
                "targetType": r.get("targetType"),
                "targetId": r.get("targetId"),
                "targetName": r.get("targetName"),
                "runMode": r.get("runMode"),
                "startedAt": r.get("startedAt"),
                "durationMs": r.get("durationMs"),
                "envName": r.get("envName"),
                "status": r.get("status"),
                "totalPass": r.get("totalPass"),
                "totalFail": r.get("totalFail"),
                "totalSkip": r.get("totalSkip"),
                "caseCount": r.get("cases").and_then(|c| c.as_array()).map(|a| a.len()).unwrap_or(0),
            })
        })
        .collect();
    Ok(Value::Array(summaries))
}

/// Load a single report's details
#[tauri::command]
pub fn load_scenario_report(app: tauri::AppHandle, id: String) -> Result<Value, String> {
    let path = reports_dir(&app)?.join(format!("{}.json", id));
    let content =
        std::fs::read_to_string(&path).map_err(|e| format!("failed to read report: {}", e))?;
    serde_json::from_str(&content).map_err(|e| format!("failed to parse report: {}", e))
}

/// Delete a single report
#[tauri::command]
pub fn delete_scenario_report(app: tauri::AppHandle, id: String) -> Result<Value, String> {
    let path = reports_dir(&app)?.join(format!("{}.json", id));
    if path.exists() {
        std::fs::remove_file(&path).map_err(|e| format!("failed to delete report: {}", e))?;
    }
    Ok(serde_json::json!({ "status": "ok" }))
}

/// Clear all reports
#[tauri::command]
pub fn clear_scenario_reports(app: tauri::AppHandle) -> Result<Value, String> {
    let dir = reports_dir(&app)?;
    if dir.exists() {
        std::fs::remove_dir_all(&dir).map_err(|e| format!("failed to clear reports: {}", e))?;
    }
    Ok(serde_json::json!({ "status": "ok" }))
}

/// Capacity trim: keep the newest MAX_KEEP entries (oldest evicted in ascending startedAt order)
fn trim_reports(dir: &PathBuf) -> Result<(), String> {
    let mut entries: Vec<(i64, PathBuf)> = Vec::new();
    for entry in std::fs::read_dir(dir).map_err(|e| e.to_string())? {
        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        let started = std::fs::read_to_string(&path)
            .ok()
            .and_then(|c| serde_json::from_str::<Value>(&c).ok())
            .and_then(|v| v.get("startedAt").and_then(|s| s.as_i64()))
            .unwrap_or_else(|| {
                entry
                    .metadata()
                    .ok()
                    .and_then(|m| m.modified().ok())
                    .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                    .map(|d| d.as_millis() as i64)
                    .unwrap_or(0)
            });
        entries.push((started, path));
    }
    if entries.len() <= MAX_KEEP {
        return Ok(());
    }
    entries.sort_by(|a, b| a.0.cmp(&b.0));
    let remove_count = entries.len() - MAX_KEEP;
    for (_, path) in entries.into_iter().take(remove_count) {
        let _ = std::fs::remove_file(path);
    }
    Ok(())
}
